File: beacon-positioning/android/app/src/main/java/com/capstone/blepos/RssiExporter.kt
package com.capstone.blepos

import org.json.JSONArray
import org.json.JSONObject
import java.io.File

/**
 * 코어 알고리즘 공용 포맷 내보내기 — iOS 와 완전히 동일한 스키마 (변경 금지).
 * CSV 헤더: timestamp,beacon_id,rssi,channel,x_true,y_true,noise_flag
 * JSON: {"meta": {...}, "samples": [{"t","beacon_id","rssi","channel"}, ...]}
 */
class RssiExporter {

    private val samples = mutableListOf<BeaconScanner.RssiSample>()

    fun add(s: BeaconScanner.RssiSample) = synchronized(samples) { samples.add(s) }

    fun toCsv(file: File) {
        synchronized(samples) {
            file.printWriter().use { w ->
                w.println("timestamp,beacon_id,rssi,channel,x_true,y_true,noise_flag")
                samples.forEach { s ->
                    w.println("${s.timestamp},${s.beaconId},${s.rssi},,,,")
                }
            }
        }
    }

    fun toJson(file: File, meta: Map<String, String> = emptyMap()) {
        synchronized(samples) {
            val arr = JSONArray()
            samples.forEach { s ->
                arr.put(JSONObject().apply {
                    put("t", s.timestamp)
                    put("beacon_id", s.beaconId)
                    put("rssi", s.rssi)
                    put("channel", JSONObject.NULL)
                })
            }
            val root = JSONObject().apply {
                put("meta", JSONObject(meta + mapOf("source" to "android", "format" to "track4-unity-v1")))
                put("samples", arr)
            }
            file.writeText(root.toString(1))
        }
    }

    fun clear() = synchronized(samples) { samples.clear() }
    val size: Int get() = synchronized(samples) { samples.size }
}

File: beacon-positioning/ios/Sources/RssiExporter.swift
import Foundation

/// Android RssiExporter.kt 와 완전히 동일한 스키마 (변경 금지).
/// CSV: timestamp,beacon_id,rssi,channel,x_true,y_true,noise_flag
final class RssiExporter {

    private var samples: [RssiSample] = []
    private let lock = NSLock()

    func add(_ s: RssiSample) {
        lock.lock(); defer { lock.unlock() }
        samples.append(s)
    }

    var count: Int {
        lock.lock(); defer { lock.unlock() }
        return samples.count
    }

    func writeCsv(to url: URL) throws {
        lock.lock(); defer { lock.unlock() }
        var out = "timestamp,beacon_id,rssi,channel,x_true,y_true,noise_flag\n"
        for s in samples {
            out += "\(s.timestamp),\(s.beaconId),\(s.rssi),,,,\n"
        }
        try out.write(to: url, atomically: true, encoding: .utf8)
    }

    func writeJson(to url: URL, meta: [String: String] = [:]) throws {
        lock.lock(); defer { lock.unlock() }
        var metaAll = meta
        metaAll["source"] = "ios"
        metaAll["format"] = "track4-unity-v1"
        let payload: [String: Any] = [
            "meta": metaAll,
            "samples": samples.map { s -> [String: Any] in
                ["t": s.timestamp, "beacon_id": s.beaconId, "rssi": s.rssi,
                 "channel": NSNull()]
            },
        ]
        let data = try JSONSerialization.data(withJSONObject: payload,
                                              options: [.prettyPrinted])
        try data.write(to: url)
    }

    func clear() {
        lock.lock(); defer { lock.unlock() }
        samples.removeAll()
    }
}
